flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerServiceImpl.java

```diff
@@ -0,0 +1,347 @@
+                    }
+                });
+    }
+
+    @Override
+    public void revokeLeadership() {
+        handleLeaderEventExecutor.execute(
+                () -> {
+                    synchronized (lock) {
+                        if (!running) {
+                            LOG.info(
+                                    "Resource manager service is not running. Ignore revoking leadership.");
+                            return;
+                        }
+
+                        LOG.info(
+                                "Resource manager service is revoked leadership with session id {}.",
+                                leaderSessionID);
+
+                        stopLeaderResourceManager();
+                    }
+                });
+    }
+
+    @Override
+    public void handleError(Exception exception) {
+        fatalErrorHandler.onFatalError(
+                new FlinkException(
+                        "Exception during leader election of resource manager occurred.",
+                        exception));
+    }
+
+    // ------------------------------------------------------------------------
+    //  Internal
+    // ------------------------------------------------------------------------
+
+    @GuardedBy("lock")
+    private void startNewLeaderResourceManager(UUID newLeaderSessionID) throws Exception {
+        stopLeaderResourceManager();
+
+        this.leaderSessionID = newLeaderSessionID;
+        this.leaderResourceManager =
+                resourceManagerFactory.createResourceManager(
+                        rmProcessContext, newLeaderSessionID, ResourceID.generate());
+
+        final ResourceManager<?> newLeaderResourceManager = this.leaderResourceManager;
+
+        previousResourceManagerTerminationFuture
+                .thenComposeAsync(
+                        (ignore) -> {
+                            synchronized (lock) {
+                                return startResourceManagerIfIsLeader(newLeaderResourceManager);
+                            }
+                        },
+                        handleLeaderEventExecutor)
+                .thenAcceptAsync(
+                        (isStillLeader) -> {
+                            if (isStillLeader) {
+                                leaderElectionService.confirmLeadership(
+                                        newLeaderSessionID, newLeaderResourceManager.getAddress());
+                            }
+                        },
+                        ioExecutor);
+    }
+
+    /**
+     * Returns a future that completes as {@code true} if the resource manager is still leader and
+     * started, and {@code false} if it's no longer leader.
+     */
+    @GuardedBy("lock")
+    private CompletableFuture<Boolean> startResourceManagerIfIsLeader(
+            ResourceManager<?> resourceManager) {
+        if (isLeader(resourceManager)) {
+            resourceManager.start();
+            forwardTerminationFuture(resourceManager);
```

Comment:
I do not find it is problem, but it is a little confusing to me that completing `ResourceManager` will also complete the `serviceTerminationFuture`, which may cause the JobManager process to exit.

Reply:
IIUC, your point is that in case the leader RM is terminated unexpectedly, the RMService can try start another RM without failing the entire process?

In fact, it's not clear to me why would a leader RM be terminated unexpectedly without losing leadership. Ideally, this should not happen. Failing strictly rather than restarting silently in such cases may help drawing attentions to such unexpected cases.

Besides, failing strictly does not introduce any regression comparing to previous behaviors. Trying to start a new RM can be considered as an future improvement if it is proven necessary.